### PR TITLE
[Draft] DLPNO-CCSD PR

### PR DESCRIPTION
**Pull request by andyj10224**
## Description
HERE IT IS!!! This is a draft of the DLPNO-CCSD PR that will be coming in the next few months. The purpose of this is for the developers and research groups to be able to run and test DLPNO-CCSD before it is officially part of the code. 

Credit to @JoseMadriaga for the derivations 
[LocalCCSD1to10.pdf](https://github.com/psi4/psi4/files/11114337/LocalCCSD1to10.pdf)

Useful References:
[Original DLPNO-CCSD Paper](https://aip.scitation.org/doi/full/10.1063/1.47735)
[Sparse Maps II Paper](https://aip.scitation.org/doi/10.1063/1.4939030)

Example Input File
```
memory 20 GB

molecule mol {
  0 1
  O    0.705    0.744    0.16
  H    -0.071    0.264    0.45
  H    1.356    0.064    -0.014
  symmetry c1
}

set {
  basis cc-pVDZ
  scf_type df
  freeze_core true
  pno_convergence normal
}
energy('dlpno-ccsd')
```

### Results (Waterclusters in TZ)
[Speedups, relative to DF-CCSD]
wc_tz_speedups

[Percent Correlation Energy Recovered, relative to DF-CCSD, all >= 99.9%]
wc_tz_percent_corr_recovered

## User API & Changelog headlines

- [x] Implement the DLPNO-CCSD algorithm

## Dev notes & details

- [ ] Feel free to use this code, it is not fully tested yet, but preliminary tests show encouraging results, and is MUCH faster than conventional CCSD
- [ ] If you benchmark my code, please post results in the thread

## Questions
- [ ] Question1

## Checklist
- [ ] Add documentation
- [ ] Add references to terms and equations
- [ ] Tests added for any new features
- [ ] [All or relevant fraction of full tests run](http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)

## Status
- [ ] Ready for review
- [ ] Ready for merge

### CODE REVIEW

File: psi4/src/psi4/dlpno/ccsd.cc [hunk 1]
```diff
@@ -0,0 +1,1151 @@
...
+    } else {
+        return S_pno_ij_mn_[ij][mn_ij];
+    }
+}
+
+void DLPNOCCSD::compute_pno_overlaps() {
+    int n_lmo_pairs = ij_to_i_j_.size();
+    int naocc = nalpha_ - nfrzc();
+
+    S_pno_ij_mn_.resize(n_lmo_pairs);
```
Comment (TiborGY) [inline psi4/src/psi4/dlpno/ccsd.cc:94]:
```suggestion
    const int n_lmo_pairs = ij_to_i_j_.size();
    S_pno_ij_mn_.resize(n_lmo_pairs);
```
`naocc` is unused

File: psi4/src/psi4/dlpno/ccsd.cc [hunk 2]
```diff
@@ -0,0 +1,1151 @@
...
+
+#pragma omp parallel for schedule(dynamic, 1)
+    for (int ij = 0; ij < n_lmo_pairs; ++ij) {
+        int i, j;
+        std::tie(i, j) = ij_to_i_j_[ij];
+
+        int npno_ij = n_pno_[ij];
+        int nlmo_ij = lmopair_to_lmos_[ij].size();
```
Comment (TiborGY) [inline psi4/src/psi4/dlpno/ccsd.cc:102]:
```suggestion
        const int npno_ij = n_pno_[ij];
        const int nlmo_ij = lmopair_to_lmos_[ij].size();
```

File: psi4/src/psi4/dlpno/ccsd.cc [hunk 3]
```diff
@@ -0,0 +1,1151 @@
...
+        if (npno_ij == 0 || i < j) continue;
+
+        S_pno_ij_mn_[ij].resize(nlmo_ij * nlmo_ij);
+
+        for (int mn_ij = 0; mn_ij < nlmo_ij * nlmo_ij; mn_ij++) {
+            int m_ij = mn_ij / nlmo_ij, n_ij = mn_ij % nlmo_ij;
+            int m = lmopair_to_lmos_[ij][m_ij], n = lmopair_to_lmos_[ij][n_ij];
+            int mn = i_j_to_ij_[m][n];
+
+            int npno_mn = n_pno_[mn];
```
Comment (TiborGY) [inline psi4/src/psi4/dlpno/ccsd.cc:113]:
```suggestion
            const int m_ij = mn_ij / nlmo_ij, n_ij = mn_ij % nlmo_ij;
            const int m = lmopair_to_lmos_[ij][m_ij], n = lmopair_to_lmos_[ij][n_ij];
            const int mn = i_j_to_ij_[m][n];

            const int npno_mn = n_pno_[mn];
```

File: psi4/src/psi4/dlpno/ccsd.cc [hunk 4]
```diff
@@ -0,0 +1,1151 @@
...
+        }
+    }
+}
+
+void DLPNOCCSD::estimate_memory() {
+    outfile->Printf("   => DLPNO-CCSD Memory Estimate <= \n\n");
+
+    int n_lmo_pairs = ij_to_i_j_.size();
+
+    size_t qvv = 0L;
```
Comment (TiborGY) [inline psi4/src/psi4/dlpno/ccsd.cc:127]:
```suggestion
    const int n_lmo_pairs = ij_to_i_j_.size();

    size_t qvv = 0;
```

File: psi4/src/psi4/dlpno/ccsd.cc [hunk 5]
```diff
@@ -0,0 +1,1151 @@
...
+#pragma omp parallel for schedule(dynamic) reduction(+ : qvv)
+    for (int ij = 0; ij < n_lmo_pairs; ij++) {
+        int i, j;
+        std::tie(i, j) = ij_to_i_j_[ij];
+
+        if (i > j) continue;
+
+        int naux_ij = lmopair_to_ribfs_[ij].size();
+        int npno_ij = n_pno_[ij];
```
Comment (TiborGY) [inline psi4/src/psi4/dlpno/ccsd.cc:136]:
```suggestion
        const int naux_ij = lmopair_to_ribfs_[ij].size();
        const int npno_ij = n_pno_[ij];
```

File: psi4/src/psi4/dlpno/ccsd.cc [hunk 6]
```diff
@@ -0,0 +1,1151 @@
...
+
+        qvv += naux_ij * npno_ij * npno_ij;
+    }
+
+    size_t qvv_memory = qvv * sizeof(double);
```
Comment (TiborGY) [inline psi4/src/psi4/dlpno/ccsd.cc:141]:
```suggestion
    const size_t qvv_memory = qvv * sizeof(double);
```

File: psi4/src/psi4/dlpno/dlpnobase.cc [hunk 7]
```diff
@@ -0,0 +1,1448 @@
...
+    print_ = options_.get_int("PRINT");
+    debug_ = options_.get_int("DEBUG");
+
+    T_CUT_PNO_ = options_.get_double("T_CUT_PNO");
+    T_CUT_PNO_DIAG_SCALE_ = options_.get_double("T_CUT_PNO_DIAG_SCALE");
+    T_CUT_DO_ = options_.get_double("T_CUT_DO");
+
+    // did the user manually change expert level options?
+    bool T_CUT_PNO_changed = options_["T_CUT_PNO"].has_changed();
+    bool T_CUT_DO_changed = options_["T_CUT_DO"].has_changed();
```
Comment (TiborGY) [inline psi4/src/psi4/dlpno/dlpnobase.cc:75]:
```suggestion
    const bool T_CUT_PNO_changed = options_["T_CUT_PNO"].has_changed();
    const bool T_CUT_DO_changed = options_["T_CUT_DO"].has_changed();
```

**Review by TiborGY — COMMENTED** [review]
Perhaps this is a bit early, but I scrolled through some of the code.
Most of the suggestions are just adding `const` to variables that are write-once, the only substantial one is where a check for DGESV failure is added.

File: psi4/src/read_options.cc [hunk 8]
```diff
@@ -2502,11 +2502,15 @@ int read_options(const std::string &name, Options &options, bool suppress_printi
         options.add_str("DLPNO_LOCAL_ORBITALS", "BOYS", "BOYS PIPEK_MEZEY");
         /*- Maximum number of iterations to determine the MP2 amplitudes. -*/
         options.add_int("DLPNO_MAXITER", 50);
+        /*- Perform DLPNO-CCSD? */
```
Comment (loriab) [inline psi4/src/read_options.cc:2505]:
```suggestion
        /*- Perform DLPNO-CCSD? !expert*/
```
user shouldn't set this, only the driver for internal routing

**Review by loriab — COMMENTED** [review]
Exciting new method! Here's a few minor changes.

File: psi4/src/psi4/dlpno/ccsd.cc [hunk 9]
```diff
@@ -0,0 +1,1151 @@
...
+    outfile->Printf("   SVD Memory/DF Memory : %8.4f \n\n", (double) qvv_svd_memory / qvv_memory);
+}
+
+SharedMatrix DLPNOCCSD::compute_Fmi(const std::vector<SharedMatrix>& tau_tilde) {
+    timer_on("Compute Fmi");
+
+    int n_lmo_pairs = ij_to_i_j_.size();
+    int naocc = nalpha_ - nfrzc();
+
+    // Equation 40, Term 1
```
[the remark this answers is not in the document]
Comment (andyj10224) [reply psi4/src/psi4/dlpno/ccsd.cc:302]: This is not referring to a paper, but the derivations by Jose Madriaga (it is referenced at the top of this page)

**Comment by andyj10224** [comment]
Just implemented some of Lori and Tibor's suggestions. I have also implemented SC-LMP2 for "weak pairs" to reduce the cost of the LCCSD computation, per the Sparse Maps II paper.

**Review by TiborGY — COMMENTED** [review]
Round 2, mostly more of the same: const correctness, a few unused variables, a docstring suggestion, and a possible suggestion to eliminate a bit of code duplication,

**Comment by davpoolechem** [comment]
One thing we will need to discuss up-front, is what to do with a DLPNO-CC Psi4NumPy implementation. I know @JonathonMisiewicz talked about this in a previous PR, and I see his point. It would be useful to have an easier-to-understand reference for DLPNO-CC somewhere, especially to help others understand the DLPNO formalism.  

A lot of the work done in this PR was previously prototyped in Python, so that gives us a start in a potential Psi4NumPy implementation.

**Pull request closed without merging**